File: tugas_Heap/src/MaxHeapData.java
import java.util.ArrayList;

/**
 * MaxHeapData class - Implementasi Max-Heap untuk DataEntry (sorted descending by id)
 * Menyediakan operasi: insert, delete, peek, dan display
 */
public class MaxHeapData {
    private ArrayList<DataEntry> heap;

    /**
     * Constructor - inisialisasi heap kosong
     */
    public MaxHeapData() {
        this.heap = new ArrayList<>();
    }

    /**
     * Operasi INSERT - menambahkan elemen ke max-heap
     * Kompleksitas: O(log n)
     */
    public void insert(DataEntry entry) {
        heap.add(entry);
        int index = heap.size() - 1;

        // Heapify-up: bandingkan dengan parent dan swap jika perlu (untuk max-heap)
        while (index > 0) {
            int parentIndex = (index - 1) / 2;
            if (heap.get(index).reverseCompareTo(heap.get(parentIndex)) < 0) {
                swap(index, parentIndex);
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    /**
     * Operasi PEEK - mendapatkan elemen terbesar (root) tanpa menghapusnya
     * Kompleksitas: O(1)
     */
    public DataEntry peek() {
        if (heap.isEmpty()) {
            return null;
        }
        return heap.get(0);
    }

    /**
     * Operasi DELETE MAX - menghapus elemen terbesar (root)
     * Kompleksitas: O(log n)
     */
    public DataEntry deleteMax() {
        if (heap.isEmpty()) {
            return null;
        }

        DataEntry max = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);

        if (!heap.isEmpty()) {
            heapifyDown(0);
        }

        return max;
    }

    /**
     * Operasi DELETE - menghapus entry spesifik dari heap
     * Kompleksitas: O(n)
     */
    public boolean delete(int id) {
        int index = findIndex(id);

        if (index == -1) {
            return false;
        }

        heap.set(index, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);

        if (index < heap.size()) {
            heapifyDown(index);
        }

        return true;
    }

    /**
     * Helper: Heapify-down - mempertahankan sifat max-heap ke bawah
     * Kompleksitas: O(log n)
     */
    private void heapifyDown(int index) {
        int largest = index;
        int leftChild = 2 * index + 1;
        int rightChild = 2 * index + 2;

        if (leftChild < heap.size() && heap.get(leftChild).reverseCompareTo(heap.get(largest)) < 0) {
            largest = leftChild;
        }

        if (rightChild < heap.size() && heap.get(rightChild).reverseCompareTo(heap.get(largest)) < 0) {
            largest = rightChild;
        }

        if (largest != index) {
            swap(index, largest);
            heapifyDown(largest);
        }
    }

    /**
     * Helper: Tukar dua elemen di heap
     */
    private void swap(int i, int j) {
        DataEntry temp = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, temp);
    }

    /**
     * Helper: Cari index entry dengan id tertentu
     */
    private int findIndex(int id) {
        for (int i = 0; i < heap.size(); i++) {
            if (heap.get(i).getId() == id) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Utility: Cek apakah heap kosong
     */
    public boolean isEmpty() {
        return heap.isEmpty();
    }

    /**
     * Utility: Dapatkan ukuran heap
     */
    public int size() {
        return heap.size();
    }

    /**
     * Utility: Dapatkan semua data dalam urutan max-heap (descending)
     */
    public ArrayList<DataEntry> getAllSorted() {
        ArrayList<DataEntry> result = new ArrayList<>();
        MaxHeapData tempHeap = new MaxHeapData();

        // Copy semua elemen
        for (DataEntry entry : heap) {
            tempHeap.insert(entry);
        }

        // Extract semua elemen dalam urutan descending
        while (!tempHeap.isEmpty()) {
            result.add(tempHeap.deleteMax());
        }

        return result;
    }

    /**
     * Utility: Display heap
     */
    public void display() {
        if (heap.isEmpty()) {
            System.out.println("Max-Heap kosong!");
            return;
        }

        ArrayList<DataEntry> sorted = getAllSorted();
        System.out.println("┌─────┬─────────────────────────────┐");
        System.out.println("│ ID  │ NAMA                        │");
        System.out.println("├─────┼─────────────────────────────┤");
        for (DataEntry entry : sorted) {
            System.out.println("│ " + entry.toString() + " │");
        }
        System.out.println("└─────┴─────────────────────────────┘");
    }

    /**
     * Utility: Get heap list (internal structure)
     */
    public ArrayList<DataEntry> getHeapList() {
        return new ArrayList<>(heap);
    }
}

File: tugas_Heap/heap_py/max_heap_data.py
from data_entry import DataEntry


class MaxHeapData:
    """Implementasi Max-Heap untuk DataEntry, diurutkan descending berdasarkan id."""

    def __init__(self):
        self.heap: list[DataEntry] = []

    def insert(self, entry: DataEntry) -> None:
        """Menambahkan elemen ke max-heap. Kompleksitas: O(log n)."""
        self.heap.append(entry)
        self._heapify_up(len(self.heap) - 1)

    def peek(self) -> DataEntry | None:
        """Mengambil elemen terbesar tanpa menghapusnya. Kompleksitas: O(1)."""
        if self.is_empty():
            return None
        return self.heap[0]

    def delete_max(self) -> DataEntry | None:
        """Menghapus elemen terbesar/root. Kompleksitas: O(log n)."""
        if self.is_empty():
            return None

        max_entry = self.heap[0]
        last_entry = self.heap.pop()

        if not self.is_empty():
            self.heap[0] = last_entry
            self._heapify_down(0)

        return max_entry

    def delete(self, id_data: int) -> bool:
        """Menghapus entry spesifik berdasarkan id. Kompleksitas: O(n)."""
        index = self._find_index(id_data)

        if index == -1:
            return False

        last_entry = self.heap.pop()

        if index < len(self.heap):
            self.heap[index] = last_entry
            self._fix_heap_after_replace(index)

        return True

    def is_empty(self) -> bool:
        return len(self.heap) == 0

    def size(self) -> int:
        return len(self.heap)

    def get_all_sorted(self) -> list[DataEntry]:
        """Mengambil semua data dalam urutan descending tanpa merusak heap asli."""
        result: list[DataEntry] = []
        temp_heap = MaxHeapData()

        for entry in self.heap:
            temp_heap.insert(entry)

        while not temp_heap.is_empty():
            result.append(temp_heap.delete_max())

        return result

    def display(self) -> None:
        if self.is_empty():
            print("Max-Heap kosong!")
            return

        sorted_entries = self.get_all_sorted()
        print("+-------+--------------------------------+")
        print("| ID    | NAMA                           |")
        print("+-------+--------------------------------+")
        for entry in sorted_entries:
            print(f"| {entry} |")
        print("+-------+--------------------------------+")

    def get_heap_list(self) -> list[DataEntry]:
        return self.heap.copy()

    def _heapify_up(self, index: int) -> None:
        while index > 0:
            parent_index = (index - 1) // 2
            if self.heap[index].reverse_compare_to(self.heap[parent_index]) < 0:
                self._swap(index, parent_index)
                index = parent_index
            else:
                break

    def _heapify_down(self, index: int) -> None:
        while True:
            largest = index
            left_child = 2 * index + 1
            right_child = 2 * index + 2

            if (
                left_child < len(self.heap)
                and self.heap[left_child].reverse_compare_to(self.heap[largest]) < 0
            ):
                largest = left_child

            if (
                right_child < len(self.heap)
                and self.heap[right_child].reverse_compare_to(self.heap[largest]) < 0
            ):
                largest = right_child

            if largest == index:
                break

            self._swap(index, largest)
            index = largest

    def _fix_heap_after_replace(self, index: int) -> None:
        parent_index = (index - 1) // 2
        if index > 0 and self.heap[index].reverse_compare_to(self.heap[parent_index]) < 0:
            self._heapify_up(index)
        else:
            self._heapify_down(index)

    def _swap(self, i: int, j: int) -> None:
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def _find_index(self, id_data: int) -> int:
        for index, entry in enumerate(self.heap):
            if entry.get_id() == id_data:
                return index
        return -1
